### PR TITLE
Feature/event list v3

### PR DESCRIPTION
**Pull request by Aurelien9Code**
This PR introduces updates to the Home page (also known as the events list). It includes final UI changes to the Event card (for now – I may revisit elements like the "favorite" button later). There's also been extensive refactoring, with class names changed and functions moved for greater consistency.

Special care was taken to ensure that everything is properly mocked so that the Preview functionality works smoothly. The "favorite" button is now functional and integrated with Firebase. **Important**: To make it work, ensure there is a `savedEvent` list in your Firebase database. This list may not exist for users created before the introduction of the event-saving feature.

Additionally, I've split what was once a shared view model for both the EventList and EventCard into two separate view models, one for each.

I’ve also added two specific images to handle image loading issues: a "No image found" placeholder for cases when Firebase fails to fetch an image, and a "Preview image" for mockups in Preview mode.

As for testing, you’ll notice there’s a lack of tests at the moment. I plan to create a level of abstraction for all Firebase-related tests using Mockito, making it easier to mock Firebase actions. Since edge cases are similar across class implementations, the idea is to consolidate these tests into a single file. Adding new tests would simply involve calling this class with specific parameters. This will be my next PR. Thanks for reviewing !

### CODE REVIEW

**Review by oskar-codes — CHANGES_REQUESTED** [review]
Overall good additions that will be muched needed for Unio 🚀 
However, several issues still need to be addressed and thus I will request changes on this pull request. For instance, running this PR locally crashes the application.

Overall, I believe a thorough architectural change is needed. Logic concerning users should not be placed in event-related files. Requests shouldn't be made to check individual properties of a user.
Also, please remove mocks and previews as much as possible. Pull-requests should be considered as clean production-ready code that can be merged without the need for further maintenance.

File: app/src/main/java/com/android/unio/ui/home/Home.kt [hunk 1]
```diff
@@ -68,16 +71,14 @@ fun EventListOverviewPreview() {
...
-      onEventClick = {})
+      eventViewModel = PreviewEventViewModel(MockEventRepository(), MockUserRepository()))
 }
 
 @Composable
 fun HomeScreen(
     navigationAction: NavigationAction,
     eventListViewModel: EventListViewModel = viewModel(factory = EventListViewModel.Factory),
-    onAddEvent: () -> Unit,
-    onEventClick: (Event) -> Unit
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/home/Home.kt:80]: Thanks for this removal. It was confusing when there was the foundation of an "add event" logic, while we had discussed that we would put a button that leads to the map.

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+  LaunchedEffect(event.uid) {
+    viewModel.isEventSavedForCurrentUser(event.uid) { saved -> isSaved = saved }
+  }
+
+  Column(
+      modifier =
+          Modifier.fillMaxWidth()
+              .padding(vertical = 8.dp)
+              .clickable(onClick = {})
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/event/EventCard.kt:98]: Perhaps it is better to simply not write the modifier if an empty callback function is put. However since this is a placeholder to add the logic later, consider adding a small comment to indicate it.

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                        .testTag("event_EventImage"),
+                contentScale = ContentScale.Crop // Crop the image to fit
+                )
+          }
+
+          Box(
+              modifier =
+                  Modifier.size(28.dp)
+                      .clip(RoundedCornerShape(14.dp))
+                      .background(Color(0xFF6200EE))
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/event/EventCard.kt:138]: Careful on hardcoded colors

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                                .testTag("event_EventMainType"),
+                        color = Color.Black,
+                        style = TextStyle(fontSize = 8.sp))
+                  }
+            }
+            Spacer(modifier = Modifier.width(6.dp))
+
+            Image(
+                painter = painterResource(id = R.drawable.clic),
+                contentDescription = null,
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/event/EventCard.kt:199]: Consider adding an actual content description!

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/event/EventCard.kt:1]: Also it might be useful to comment what each composable structure is for. Since there are a lot of compostables embedded in each other, it is not easy to immediately identify what the composables correspond to on the event card.

**Comment by Zafouche** [comment]
Otherwise lots of work done, thank you!

File: app/src/main/java/com/android/unio/MainActivity.kt [hunk 5]
```diff
@@ -94,11 +94,7 @@ fun UnioApp() {
     }
     navigation(startDestination = Screen.HOME, route = Route.HOME) {
       composable(Screen.HOME) {
-        HomeScreen(
-            navigationActions,
-            EventListViewModel(EventRepositoryFirestore(db)),
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/MainActivity.kt:97]: thanks for the tips !

File: app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt [hunk 6]
```diff
@@ -42,6 +47,85 @@ class UserRepositoryFirestore(private val db: FirebaseFirestore) : UserRepositor
...
+          Log.d("Firestore", "Event successfully removed from saved events for user: $userUid")
+          onSuccess()
+        }
+        .addOnFailureListener { e ->
+          Log.w("Firestore", "Error removing event for user: $userUid", e)
+          onFailure(e)
+        }
+  }
+
+  override fun isEventSaved(userUid: String, eventUid: String, onResult: (Boolean) -> Unit) {
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt:93]: In order not to have to create a getter (fun getUids(): List<String> = _uids) in the FirestoreReferenceList, I think I will instead create a generic contains<T> function inside the ReferenceList interface ; )

File: app/src/main/java/com/android/unio/model/event/EventViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    currentUser?.uid?.let { userUid ->
+      userRepository.listenToSavedEvents(userUid, onSavedEventsChanged)
+    }
+        ?: run {
+          Log.w("Firestore", "No user logged in, unable to listen to saved events")
+          onSavedEventsChanged(emptyList())
+        }
+  }
+
+  fun saveEventForCurrentUser(
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/event/EventViewModel.kt:82]: You are completely right !

File: app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt [hunk 8]
```diff
@@ -1,15 +1,20 @@
...
+import com.google.firebase.firestore.FieldValue
 import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.firestore
 
 class UserRepositoryFirestore(private val db: FirebaseFirestore) : UserRepository {
 
-  override fun init(onSuccess: () -> Unit) {
+  override fun init(
+      onSuccess: () -> Unit
+  ) { // repository is only considered "initialized" when a user is authenticated
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt:17]: hmmm then I don't understand the purpose of the init function.. ?

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+          Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
+            Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.weight(1f)) {
+              Text(
+                  modifier =
+                      Modifier.padding(vertical = 1.dp, horizontal = 4.dp)
+                          .testTag("event_EventLocation")
+                          .wrapContentWidth(), // Make sure the text only takes as much space as
+                  // needed
+                  text = event.location.name,
+                  style = MaterialTheme.typography.bodySmall,
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:217]: Good catch !

File: app/src/test/java/com/android/unio/model/event/EventViewModelTest.kt
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/event/EventViewModelTest.kt:1]: Thanks !!! ; )

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+          }
+
+          Text(
+              modifier =
+                  Modifier.padding(vertical = 1.dp, horizontal = 4.dp)
+                      .testTag("event_EventCatchyDescription")
+                      .wrapContentWidth(),
+              text = event.catchyDescription,
+              style = MaterialTheme.typography.bodySmall,
+              color = Color.Black)
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:257]: I would disagree... For the EventCard item, the black is a good color I think for the text informations (if you don't think so, I'll be very glad to change it for another color !)
But I we say black is the best color here, then adding the definition of the Black color in theme/Color.kt is absurd, it already exists in the Color package of Kotlin and thus it is not hardcoded

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt (hunk 4 above)
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:201]: thanks ! Good catch

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt (hunk 3 above)
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:140]: Ok this one's a good catch ! Thanks

File: app/src/androidTest/java/com/android/unio/screen/EventListOverviewTest.kt [hunk 11]
```diff
@@ -47,31 +48,25 @@ class EventListOverviewTest {
    * 'All' tab exists and can be clicked, and verifies the underlying bar's presence when switching
    * tabs.
    */
-  @Test
+  /*@Test
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/androidTest/java/com/android/unio/screen/EventListOverviewTest.kt:51]: Good catch ! thanks

File: app/src/main/java/com/android/unio/model/event/Event.kt [hunk 12]
```diff
@@ -37,3 +46,15 @@ data class Event(
...
+
+enum class EventType(val color: Color, val text: String) {
+  FESTIVAL(eventTypeFestival, "festival"),
+  APERITIF(eventTypeAperitif, "aperitif"),
+  NIGHT_PARTY(eventTypeNightParty, "night party"),
+  JAM(eventTypeJam, "jam"),
+  NETWORKING(eventTypeNetworking, "networking"),
+  SPORT(eventTypeSport, "sport"),
+  TRIP(eventTypeTrip, "trip"),
+  OTHER(eventTypeOther, "other")
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/event/Event.kt:58]: I am not sure to understand but I don't think so ! 
I think in order to preserve a color identification of the type of Event (so that we can understand which event is a festival in the blink of an eye :), it should be the same when going to light/dark mode

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt (hunk 2 above)
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:102]: That's sad because I loved doing that to remember to implement it haha but I think you are right !

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+import com.android.unio.utils.EventUtils.addAlphaToColor
+import com.android.unio.utils.EventUtils.formatTimestamp
+import com.google.firebase.Timestamp
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+
+@Preview(showBackground = true)
+@Composable
+fun EventCardPreview() {
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:73]: You are right !

File: app/src/main/java/com/android/unio/model/user/UserRepository.kt [hunk 14]
```diff
@@ -1,11 +1,121 @@
...
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  )
+
+  fun isEventSaved(userUid: String, eventUid: String, onResult: (Boolean) -> Unit)
+
   fun updateUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit)
 }
+
+class MockUserRepository : UserRepository {
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/user/UserRepository.kt:17]: You are right ! thanks

[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/user/UserRepository.kt:17]: Thanks !

File: app/src/main/java/com/android/unio/ui/association/AssociationProfile.kt [hunk 15]
```diff
@@ -259,15 +260,16 @@ fun UserCard(context: Context) {
...
   Column(
       modifier = Modifier.padding(horizontal = 28.dp),
       horizontalAlignment = Alignment.CenterHorizontally) {
         Box(modifier = Modifier.testTag("AssociationEventCard")) {
           EventCard(
               event =
-                  Event(
-                      organisers = MockReferenceList(),
-                      taggedAssociations = MockReferenceList())) {}
+                  Event(organisers = MockReferenceList(), taggedAssociations = MockReferenceList()),
```
[the remark this answers is not in the document]
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/association/AssociationProfile.kt:269]: Good catch !

**Comment by Aurelien9Code** [comment]
Heyyy ! I wanted to particularly thank you both for your review. It helped me understand a lot of things. 🚀

I know the tests won't work, I'll do them in the same PR later, but as all the changes you asked for are completed I asked you both for reviewing ;)

File: app/src/androidTest/java/com/android/unio/ui/BottomNavigationTest.kt [hunk 16]
```diff
@@ -26,7 +26,7 @@ class BottomNavigationTest {
 
   @Test
   fun testBottomNavigationMenuDisplayed() {
-    composeTestRule.setContent { HomeScreen(navigationAction, onAddEvent = {}, onEventClick = {}) }
+    composeTestRule.setContent { HomeScreen(navigationAction) }
```
[the remark this answers is not in the document]
Comment (oskar-codes) [reply app/src/androidTest/java/com/android/unio/ui/BottomNavigationTest.kt:29]: Indeed, I agree with you. Generally, I believe default parameters can cause issues, so removing factories from the default parameters might be something we need to consider next meeting.

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/event/EventCard.kt:1]: Yep ! You are completely right

File: app/src/main/java/com/android/unio/model/user/UserRepository.kt [hunk 17]
```diff
@@ -1,121 +1,11 @@
...
-      onSuccess: () -> Unit,
-      onFailure: (Exception) -> Unit
-  )
-
-  fun isEventSaved(userUid: String, eventUid: String, onResult: (Boolean) -> Unit)
-
   fun updateUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit)
 }
-
-class MockUserRepository : UserRepository {
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/model/user/UserRepository.kt]: Good call on removing this. I don't think it should have been in the same file as the actual interface in the first place.

File: app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt
Comment (Zafouche) [inline app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt:1]: What's the reasoning in removing all these function from the repository, and writing functions in the view model that do not use the repository? That breaks the meaning of a view model, which is supposed to be the middle-man between the database et the app.

File: app/src/main/java/com/android/unio/ui/association/AssociationProfile.kt [hunk 18]
```diff
@@ -257,16 +258,18 @@ fun UserCard(context: Context) {
...
       horizontalAlignment = Alignment.CenterHorizontally) {
         Box(modifier = Modifier.testTag("AssociationEventCard")) {
           EventCard(
               event =
-                  Event(organisers = MockReferenceList(), taggedAssociations = MockReferenceList()),
-              viewModel = eventViewModel)
+                  Event(
+                      organisers = Association.emptyFirestoreReferenceList(),
+                      taggedAssociations = Association.emptyFirestoreReferenceList()),
+              userViewModel = userViewModel)
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/association/AssociationProfile.kt:272]: Why use a "User" view model to fetch "Events"?

File: app/src/main/java/com/android/unio/ui/event/EventCard.kt [hunk 19]
```diff
@@ -33,118 +34,83 @@ import androidx.compose.ui.draw.clip
...
               .background(secondaryDark)) {
+
+        // Event image section, displays the main event image or a placeholder if the URL is invalid
+
         Box(modifier = Modifier.fillMaxWidth().height(100.dp)) {
-          // Fallback to a local placeholder image in case of an invalid URI
+          // fallback to a local placeholder image in case of an invalid URI
           val imageUrl = event.image.takeIf { it.isNotEmpty() }?.toUri()
 
-          if (LocalInspectionMode.current) { // preview mode
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/event/EventCard.kt]: Thanks for removing the preview logic, as it does not have its place in the app functionality

File: app/src/main/java/com/android/unio/ui/home/Home.kt [hunk 20]
```diff
@@ -197,7 +175,7 @@ fun HomeScreen(
             LazyColumn(
                 contentPadding = PaddingValues(vertical = 8.dp),
                 modifier = Modifier.fillMaxSize().padding(horizontal = 32.dp)) {
-                  items(events) { event -> EventCard(event = event, viewModel = eventViewModel) }
+                  items(events) { event -> EventCard(event = event, userViewModel = userViewModel) }
```
Comment (Zafouche) [inline app/src/main/java/com/android/unio/ui/home/Home.kt:178]: Same thing as before, why use a user view model to fetch events? This seems odd.

File: app/src/test/java/com/android/unio/model/event/EventViewModelTest.kt [hunk 21]
```diff
@@ -95,3 +97,20 @@ class EventViewModelTest {
 
   class UnsupportedViewModel : ViewModel()
 }
+
+class MockReferenceList<T>(elements: List<T> = emptyList()) : ReferenceList<T> {
```
Comment (Zafouche) [inline app/src/test/java/com/android/unio/model/event/EventViewModelTest.kt:101]: Consider injecting functionality into the existing repository that you mock instead of making a mock repository every time. It is harder to figure out but ultimately less work in the long run. I can however also agree that making a mock repository can be easier to figure out

**Review by Zafouche — COMMENTED** [review]
Good small additions that take into account our previous reviews, thanks!

However, as you can see in my comments under, I have trouble understanding the choice of refactoring the repositories and view models, and especially using a view model to "fetch different kind of data" than what it's intended for.

File: app/src/main/java/com/android/unio/model/user/UserRepository.kt (hunk 17 above)
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/user/UserRepository.kt]: You are right !

File: app/src/main/java/com/android/unio/ui/association/AssociationProfile.kt (hunk 18 above)
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/ui/association/AssociationProfile.kt:272]: Because the saved events is something that inherently belongs to the User. You can see Oskar explanation above

File: app/src/test/java/com/android/unio/model/event/EventViewModelTest.kt (hunk 21 above)
Comment (Aurelien9Code) [reply app/src/test/java/com/android/unio/model/event/EventViewModelTest.kt:101]: Sure ! Done in #123 ! : )

File: app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt
Comment (Aurelien9Code) [reply app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt:1]: Because it is not firestore functions. It is operations that should be done locally on the pre-definied user. 
And it should not be done on the UI, so the view model is fine don't you agree ?

Comment (Zafouche) [reply app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt:1]: So for saveEvent for instance, it is only saved locally and never in the server?

File: app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (oskar-codes) [reply app/src/main/java/com/android/unio/model/user/UserRepositoryFirestore.kt]: sure you are the god of this world

**Review by oskar-codes — APPROVED** [review]
The previous 17 commits are changes reviewed by myself and @Aurelien9Code. They address issues noticed in my review and @Zafouche's review, along with other issues that emerged when merging. The resulting code should be ready to merge, so I will approve the pull request.

**Review by Zafouche — APPROVED** [review]
I supervised while @Aurelien9Code and @oskar-codes were doing the last changes, LGTM.
